src/Propel/Common/Config/ConfigurationManager.php

```diff
@@ -136,43 +137,30 @@ public function getConnectionParametersArray($section = 'runtime')
      * This method also looks for a '.dist' configuration file and loads it.
      *
      * @param string|null $fileName Configuration file name or directory in which resides the configuration file.
-     * @param array|null $extraConf Array of configuration properties, to be merged with those loaded from file.
+     * @param array $extraConf Array of configuration properties, to be merged with those loaded from file.
      *
      * @return void
      */
-    protected function load($fileName, $extraConf)
+    protected function load(?string $fileName, array $extraConf = []): void
```

Comment:
@mringler 
modified your code to match the original code.
```php
protected function load(?string $path, array $extraConf = []): void
{
    $this->config = $this->loadConfig($path, $extraConf);
}

protected function loadConfig(?string $path, array $extraConf = []): array
{
    if ($path === self::CONFIG_FILE_NAME && is_file($path)) {
        throw new \LogicException('You must call one of in() or append() methods before iterating over a Finder.');
    }
    if (!$path || ($path === self::CONFIG_FILE_NAME && !is_dir($path))) {
        $path = getcwd();
    }
    if (is_dir($path)) {
        $fileName = $this->getConfigFileNameFromDirectory($path);
        if (!$fileName) {
            return $extraConf;
        }
    } else {
        $fileName = $path;
    }

    $distFileConfig = $this->loadFile($fileName . '.dist');
    $fileConfig = $this->loadFile($fileName);
    return array_replace_recursive($distFileConfig, $fileConfig, $extraConf);
}

protected function getConfigFileNameFromDirectory(string $sourceDir): ?string
{
    $dirs[] = $sourceDir;
    if (is_dir($sourceDir . '/conf')) {
        $dirs[] = $sourceDir . '/conf';
    }
    if (is_dir($sourceDir . '/config')) {
        $dirs[] = $sourceDir . '/config';
    }

    $files = $this->getFiles($dirs, self::CONFIG_FILE_NAME);
    if(!$files) {
        $files = $this->getFiles($dirs, self::CONFIG_FILE_NAME, true);
    }

    return ($files) ? current($files)->getPathName() : null;
}
```

I think this branch is a bug:
```php
    if ($path === self::CONFIG_FILE_NAME && is_file($path)) {
        throw new \LogicException('You must call one of in() or append() methods before iterating over a Finder.');
    }
```
so is this condition:
```php
$path === self::CONFIG_FILE_NAME && !is_dir($path)
```

Also, when `getConfigFileNameFromDirectory` returns *.dist, loadConfig might read a file with `.dist.dist` extensions.

Can I fix these bugs?

Reply:
Why do we need to handle the case where `$path` is 'propel', and there is a regular file with that name in the current directory any different than when `$path` matches a differently named file? I don't think we do.

In any way, structures like this:
```php
    if (!$path || ($path === self::CONFIG_FILE_NAME && !is_dir($path))) {
        $path = getcwd();
    }
    if (is_dir($path)) {
```
where you partially fix a variable, and then use it as guard - that needs to be avoided like the plague.

Reply:
@mringler what do you mean by "partially fix a variable, and then use it as guard"?
does it mean that this
```php
if (!$path || is_dir($path)) {
    if (!$path) {
        $path = getcwd();
    }
```
is better than this?
```php
if (!$path) {
    $path = getcwd();
}
if (is_dir($path)) {
```

Reply:
maybe `getcwd()` should be in `load`, so loadConfig accepts only a string `$path`

Reply:
In your example, it does not really matter, I think both is fine. It's the complex condition around `self::CONFIG_FILE_NAME` that makes a difference for me (and maybe this is all just subjective).

But if you compare them:
```php
    if (!$path || ($path === self::CONFIG_FILE_NAME && !is_dir($path))) {
        $path = getcwd();
    }
    if (is_dir($path)) {
```
the first condition is always evaluated, and the result of the second condition is changed by the result of the first.

In contrast to:
```php
    if (!$path || is_dir($path) || $path === self::CONFIG_FILE_NAME) {
        if(!is_dir($path) || $path === self::CONFIG_FILE_NAME) {
            $path = getcwd();
        }
```
I can see directly which values satisfy the first condition, and the second condition is not influenced by the result of the first (apart from not running at all if the first one fails).

And I think it is this removed inter-dependency that makes the second one much easier to comprehend.
Does that make sense? Do you (dis)agree?

I probably wouldn't feel so strongly about it, if I hadn't just untangled the old version of this method, which I found exhausting, particularly because it was full of those inter-dependencies between conditions.

Reply:
Oh, I don't really understand what you want to achieve with your changes. Would you mind explaining it to me?

Reply:
@mringler I thought `loadConfig` should be for searching and loading behavior, and default value handling should be a concern of others (load or __construct).
then `loadConfig` is easier to override and to test.

Reply:
also `load` and `loadConfig` sounds almost the same, so I think I should rename one of them or inline `load`.

Reply:
this is what I meant https://github.com/propelorm/Propel2/pull/1753/commits/c0e66a44f430da93af40e50a5057e59f861087dd